gcc/rust/checks/errors/borrowck/rust-borrow-checker.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-borrow-checker.h"
+#include "rust-function-collector.h"
+#include "rust-bir-builder.h"
+#include "rust-bir-dump.h"
+
+namespace Rust {
+namespace HIR {
+
+void
+mkdir_wrapped (const std::string &dirname)
+{
+  int ret;
+#ifdef _WIN32
+  ret = _mkdir (dirname.c_str ());
+#elif unix
+  ret = mkdir (dirname.c_str (), 0775);
+#elif __APPLE__
+  ret = mkdir (dirname.c_str (), 0775);
```

Comment:
Actually not really, since failing to open a file will handle that error. I thought about it but I don't care about failure for example if it already exists - which might get annoying to track for different platforms.

> Also, should that function live in rust-borrow-checker.cc and not in rust-bir-dump.cc?
From my original point of view, the caller (rust-borrow-checker.cc) handles the saving of the dump. But now thinking about it, it might be better there, cause it is not interesting for borrowchecking.

Reply:
It would be probably better to use here the mkdir from `gcc/system.h` (as Thomas suggested on Zulip) then the ifdefs which I took from some other part of gcc. One problem there is that `mkdir` symbol is deprecated on windows and `_mkdir` should be used. But that should probably be handled for the whole gcc in the system header once that becomes an issue.s Regarding the error handling, we need to ignore error when directory exists, but I found out that errno value macro EEXISTS is defined in C++11 (not just POSIX), so we should be safe there.

Regarding the location, I will need the mkdir also in other parts then bir dump (facts dump), so I think it should stay in `rust-borrow-checker.cc`

@CohenArthur ?

Reply:
sounds like a good explanation :) I'm okay with all of it and will mark this as resolved